tracer/test/Datadog.Trace.Tests/Propagators/W3CTraceContextPropagatorTests.cs

```diff
@@ -23,204 +25,391 @@ static W3CTraceContextPropagatorTests()
+                traceId,
+                spanId,
+                sampled,
+                rawTraceId,
+                rawParentId);
 
-            // Extract sampling from trace context
-            var newContext = new SpanContext(null, new TraceContext(null), null, traceId, spanId);
-            var newHeaders = new Mock<IHeadersCollection>();
-            W3CPropagator.Inject(newContext, newHeaders.Object, (carrier, name, value) => carrier.Set(name, value));
-            newHeaders.Verify(h => h.Set("traceparent", "00-000000000000000000000000075bcd15-000000003ade68b1-00"), Times.Once());
-            newHeaders.VerifyNoOtherCalls();
+            W3CTraceContextPropagator.TryParseTraceParent(header, out var traceParent).Should().BeTrue();
 
-            var traceContextSamplingField = typeof(TraceContext).GetField("_samplingPriority", BindingFlags.Instance | BindingFlags.NonPublic);
-            traceContextSamplingField.SetValue(newContext.TraceContext, SamplingPriorityValues.UserKeep);
-            newHeaders = new Mock<IHeadersCollection>();
-            W3CPropagator.Inject(newContext, newHeaders.Object, (carrier, name, value) => carrier.Set(name, value));
-            newHeaders.Verify(h => h.Set("traceparent", "00-000000000000000000000000075bcd15-000000003ade68b1-01"), Times.Once());
-            newHeaders.VerifyNoOtherCalls();
+            traceParent.Should().BeEquivalentTo(expected);
+        }
+
+        // "{version:2}-{traceid:32}-{parentid:16}-{traceflags:2}
+        [Theory]
+        [InlineData(null)]                                                           // null
+        [InlineData("")]                                                             // empty
+        [InlineData(" ")]                                                            // whitespace
+        [InlineData("00-000000000000000000000000075bcd15-000000003ade68b1-0")]       // too short (length => 54)
+        [InlineData("000-00000000000000000000000075bcd15-000000003ade68b1-00")]      // wrong hyphen location (2 => 3)
+        [InlineData("00-000000000000000000000000075bcd150-00000003ade68b1-00")]      // wrong hyphen location (35 => 36)
+        [InlineData("00-000000000000000000000000075bcd15-000000003ade68b-100")]      // wrong hyphen location (52 => 51)
+        [InlineData("ff-000000000000000000000000075bcd1z-000000003ade68b1-00")]      // bad version value (ff)
+        [InlineData("xz-000000000000000000000000075bcd1z-000000003ade68b1-00")]      // bad version value (xz)
+        [InlineData("00-000000000000000000000000075bcd1z-000000003ade68b1-00")]      // bad trace id value ("z" in hex)
+        [InlineData("00-000000000000000000000000075bcd15-000000003ade68bx-00")]      // bad parent id value ("x" in hex)
+        [InlineData("00-000000000000000000000000075bcd15-000000003ade68b1-00-1234")] // do NOT allow more data after trace-flags if the version is "00"
+        public void TryParseTraceParent_Invalid(string header)
```

Comment:
Shall we add a "01" version example as well? It's not yet covered by the code, but should we be able to parse more than version 0?

Reply:
There is a `01` and a `02` version in the `TryParseTraceParent()` test above. The header is still parsed in that case it's considered valid.

Reply:
Ok, I thought we shouldn't handle yet future versions, see my other comment. Which was wrong so this other comment is also invalid. Thanks a lot

Reply:
Yeah, this is confusing. I thought the same thing when I read the W3C spec for `traceparent`, but Zach pointed out there is a separate section further down in the document, [3.2.4 Versioning of traceparent](https://www.w3.org/TR/trace-context/#versioning-of-traceparent), which covers this.